paz/models/detection/frustum_pointnet.py

```diff
@@ -0,0 +1,370 @@
+        mask: TF tensor in shape (B,N) of 0 (not pick) or 1 (pick)
+        npoints: int scalar, maximum number of points to keep (default: 512)
+    Output:
+        object_pc: TF tensor in shape (B,npoint,C)
+        indices: TF int tensor in shape (B,npoint,2)
+    """
+
+    def mask_to_indices(mask):
+        indices = np.zeros((mask.shape[0], npoints, 2), dtype=np.int32)
+        for i in range(mask.shape[0]):
+            pos_indices = np.where(mask[i, :] > 0.5)[0]
+            # skip cases when pos_indices is empty
+            if len(pos_indices) > 0:
+                if len(pos_indices) > npoints:
+                    choice = np.random.choice(len(pos_indices),
+                                              npoints, replace=False)
+                else:
+                    choice = np.random.choice(len(pos_indices),
+                                              npoints - len(pos_indices),
+                                              replace=True)
+                    choice = np.concatenate((np.arange(len(pos_indices)),
+                                             choice))
+                np.random.shuffle(choice)
+                indices[i, :, 1] = pos_indices[choice]
+            indices[i, :, 0] = i
+        return indices
+
+    indices = tf.py_function(mask_to_indices, [mask], tf.int32)
+    object_pc = tf.gather_nd(point_cloud, indices)
+    return object_pc, indices
+
+def PointCloudTranslation(point_cloud, logits, IntermediateOutputs,
+                          NUM_OBJECT_POINT=512, xyz_only=True):
+    """ Select point cloud with predicted 3D mask,
+    translate coordinates to the masked points centroid.
+
+    Used from Frustum PointNet
+
+    Input:
+        point_cloud: TF tensor in shape (B,N,C)
+        logits: TF tensor in shape (B,N,2)
+        IntermediateOutputs: dict
+        xyz_only: boolean, if True only return XYZ channels
+    Output:
+        object_point_cloud: TF tensor in shape (B,M,3)
+            for simplicity we only keep XYZ here
+            M = NUM_OBJECT_POINT as a hyper-parameter
+        mask_xyz_mean: TF tensor in shape (B,3)
+    """
+    batch_size = point_cloud.get_shape()[0]
+    num_point = point_cloud.get_shape()[1]
+    mask = tf.slice(logits, [0, 0, 0], [-1, -1, 1]) < \
+           tf.slice(logits, [0, 0, 1], [-1, -1, 1])
+    mask = tf.cast(mask, dtype=tf.float32)  # BxNx1
+    mask_count = tf.tile(tf.math.reduce_sum(mask, axis=1, keepdims=True),
+                         [1, 1, 3])  # Bx1x3
+    point_cloud_xyz = tf.slice(point_cloud, [0, 0, 0], [-1, -1, 3])  # BxNx3
+    mask_xyz_mean = tf.math.reduce_sum(
+        tf.tile(mask, [1, 1, 3]) * point_cloud_xyz, axis=1,
+        keepdims=True)  # Bx1x3
+    mask = tf.squeeze(mask, axis=[2])  # BxN
+    IntermediateOutputs['mask'] = mask
+    mask_xyz_mean = mask_xyz_mean / tf.maximum(mask_count, 1)  # Bx1x3
+
+    # Translate to masked points' centroid
+    point_cloud_xyz_stage1 = point_cloud_xyz - \
```

Comment:
please remove the use of '\' instead put everything inside a parenthesis (point_cloud_xys -
 tf.tile...)